### PR TITLE
feat(github): Implement GraphQL datasource utilities

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Implement `GithubGraphqlDatasource` class that performs paginated fetching of releases or tags (using corresponding `GithubGraphqlDatasourceAdapter` instance)
- Wrap its usage to `queryTags` and `queryReleases` functions for ease of use

This implementation is self-contained, i.e. it doesn't use types from previous implementation which will be deleted further.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
1. It doesn't really do anything as-is, right?
2. Any way to test it and be confident it works?
3. Is the use of name `datasource.ts` intentional? Or better to avoid reusing the name here in util?

**Comment by zharinov** [comment]
> 1. It doesn't really do anything as-is, right?

Right

> 2. Any way to test it and be confident it works?

You can call `await queryTags(config, this.http)` or `await queryReleases(config, this.http)` inside `getReleases` method in `tags` or `releases` datasource.

> 3. Is the use of name `datasource.ts` intentional? Or better to avoid reusing the name here in util?

It's intentional, though can be quite confusing. Any ideas for better naming?

**Comment by zharinov** [comment]
This can be enabled gradually, e.g. starting from `github-tags` and extending to `github-releases` and then changelog fetching

File: lib/util/github/graphql/datasource-helper.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  GithubPackageConfig,
+  RawQueryResponse,
+} from './types';
+
+function isSomethingWentWrongMessage(msg: string): boolean {
+  return msg.startsWith('Something went wrong while executing your query.');
+}
+
+/**
+ * See: #16343
```
Comment (viceice) [inline lib/util/github/graphql/datasource-helper.ts:25]: please add some more description

File: lib/util/github/graphql/datasource-helper.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      body: { query, variables },
+    };
+  }
+
+  private async doRawQuery(): Promise<
+    RawQueryResponse<GithubGraphqlPayload<GraphqlItem>>
+  > {
+    const requestOptions = this.getRawQueryOptions();
+
+    type HttpBody = GithubGraphqlResponse<
```
Comment (viceice) [inline lib/util/github/graphql/datasource-helper.ts:121]: double nested type? maybe on wrong type?

Comment (zharinov) [reply lib/util/github/graphql/datasource-helper.ts]: Outer type is from upper-level `http` types, inner one is specific to GraphQL datasource

**Review by viceice — APPROVED** [review]
pretty complicated 🤷‍♂️

**Comment by zharinov** [comment]
Did my best to separate concerns, though

**Pull request merged**